File: kotlin-java/src/main/kotlin/programmers/lv2/배달/배달.kt
package programmers.lv2.배달

import java.util.PriorityQueue

data class Edge(val to: Int, val cost: Int)

fun solution(N: Int, road: Array<IntArray>, k: Int): Int {
    val pq = PriorityQueue<Edge>(compareBy { it.cost })

    val graph = Array(N + 1) { mutableListOf<Edge>() }
    road.forEach {
        graph[it[0]].add(Edge(it[1], it[2]))
        graph[it[1]].add(Edge(it[0], it[2]))
    }

    val distance = IntArray(N + 1) { Int.MAX_VALUE }

    pq.add(Edge(1, 0))
    distance[1] = 0

    while (pq.isNotEmpty()) {
        val cur = pq.poll()

        graph[cur.to].forEach { next ->
            val nCost = distance[cur.to] + next.cost
            if (nCost < distance[next.to]) {
                distance[next.to] = nCost
                pq.offer(Edge(next.to, nCost))
            }
        }
    }
    return distance.count { it <= k }
}

File: Python/programmers/lv2/배달.py
import heapq


class Edge:
    def __init__(self, to, cost):
        self.to = to
        self.cost = cost

    def __lt__(self, other):
        return self.cost < other.cost


def solution(N, road, K):
    graph = [[] for _ in range(N + 1)]
    for (v1, v2, cost) in road:
        graph[v1].append(Edge(v2, cost))
        graph[v2].append(Edge(v1, cost))

    distance = [987654321] * (N + 1)
    distance[1] = 0

    priority_queue = []
    heapq.heappush(priority_queue, Edge(1, 0))

    while priority_queue:
        cur = heapq.heappop(priority_queue)
        for next in graph[cur.to]:
            new_cost = distance[cur.to] + next.cost
            if new_cost < distance[next.to]:
                distance[next.to] = new_cost
                heapq.heappush(priority_queue, Edge(next.to, new_cost))

    return len([d for d in distance if d <= K])
